### PR TITLE
Add `StreamKeyType` class and replace string literals with constants

### PR DESCRIPTION
**Pull request by anoadragon453**
This PR replaces all the instances of a hard-coded stream key string literal with a new constant type, `StreamKeyType`. No more typos!

Stream keys in unit tests were not touched, as it's useful to have them as verification (i.e. if a `StreamKeyType` constant is accidentally changed).

Recommended to review commit-by-commit.

### CODE REVIEW

File: synapse/types.py [hunk 1]
```diff
@@ -625,6 +625,22 @@ async def to_string(self, store: "DataStore") -> str:
             return "s%d" % (self.stream,)
 
 
+class StreamKeyType:
+    """Known stream types.
+
+    A stream is a list of entities ordered by an incrementing "stream token".
+    """
```
Comment (richvdh) [inline synapse/types.py:632]: a quick driveby: we should probably give more context around where these "streams" are used. Is it somethign to do with replication, or the database or is it entirely internal? If the latter, where is it used? (honestly, I've entirely paged this out. Sorry!)

**Review by richvdh — COMMENTED** [review]
I may not be able to remember what this does, but I do remember it being a thing I very much wanted, so yay!

File: synapse/types.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/types.py:632]: They're entirely internal, but quite central to Synapse - hence the slightly vague description here. One could write a lot about streams! And the documentation may be a better place for it.

Stream contents are stored in the database, and each stream is its own table with a stream token (an incrementing integer), and then other columns related to the stream type. Different internal systems processing the stream can then easily determine to what point in the stream they've processed up to. As an example, the sync token contains a list of stream tokens separated by underscores. This helps indicate to Synapse at which point in each stream that client has information about.

Another example is when sending events to application services. A token per stream type per appservice is recorded in order to determine what entities (PDUs, EDUs) still need to be pushed to that appservice (any stream entries with a token greater than the saved token).

I would like to document streams in greater detail... but I might do that in a separate area and keep this docstring on just the types light.

File: synapse/types.py [hunk 2]
```diff
@@ -625,6 +625,22 @@ async def to_string(self, store: "DataStore") -> str:
...
+
+    ROOM = "room_key"
+    PRESENCE = "presence_key"
+    TYPING = "typing_key"
+    RECEIPT = "receipt_key"
+    ACCOUNT_DATA = "account_data_key"
+    PUSH_RULES = "push_rules_key"
+    TO_DEVICE = "to_device_key"
+    DEVICE_LIST = "device_list_key"
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/types.py:641]: Thanks for the context! Yes, it looks like `StrEnum` has the drawback of interacting poorly with our string interning code.

Seeing as these types are fairly pervasive throughout the codebase, I would advocate for `Final` just to avoid hard-to-debug surprises.

File: synapse/types.py (hunk 1 above)
Comment (richvdh) [reply synapse/types.py:633]: well, I won't block this PR for it, but I would really love to see a few more words here.

**Review by H-Shay — APPROVED** [review]
LGTM!
